### PR TITLE
Adds a reader class as a wrapper to the decoding methods

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description
Implements a simple reader class for users to use the decoding methods to iterate an encoded IR stream. It also provides a way to create a generator with a customized query. The overall structure and functionality should be similar to the reader implemented in the current CLP Python logging library.

# Validation performed
1. All the unit tests passed (including newly added ones)
2. All the linters passed.

### CODE REVIEW

File: src/clp_ffi_py/ir/decoding_methods.cpp [hunk 1]
```diff
@@ -224,5 +224,75 @@ auto decode_next_log_event(PyObject* Py_UNUSED(self), PyObject* args, PyObject*
             is_query_given ? py_reinterpret_cast<PyQuery>(query) : nullptr
     );
 }
+
+auto skip_to_time(PyObject* Py_UNUSED(self), PyObject* args, PyObject* keywords) -> PyObject* {
```
Comment (davidlion) [inline src/clp_ffi_py/ir/decoding_methods.cpp]: This method appears to duplicate a significant amount of code from `decode`. I think there should be a way to refactor the common code of these two methods.

File: clp_ffi_py/readers.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    generator with a customized search query.
+
+    :param istream: Input stream that contains encoded CLP IR.
+    :param decoder_buffer_size: Initial size of the decoder buffer.
+    :param enable_compression: A flag indicating whether the istream is
+    compressed using `zstd`.
+    """
+
+    def __init__(
+        self, istream: IO[bytes], decoder_buffer_size: int = 4096, enable_compression: bool = True
```
Comment (davidlion) [inline clp_ffi_py/readers.py]: Lets set the initial size of the buffer to something more practical. The initial value I used is pretty small relative to the logs we will process and cause a lot of reads.

File: tests/test_readers.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    def setUp(self) -> None:
+        self.enable_compression = False
+        self.has_query = True
+        self.num_test_iterations = 10
+        super().setUp()
+
+
+class TestCaseReaderTimeRangeQueryZst(TestCaseReaderTimeRangeQueryBase):
+    """
+    Tests stream reader against zst compressed IR stream with the query that
```
Comment (davidlion) [inline tests/test_readers.py]: zstd

File: tests/test_readers.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    def setUp(self) -> None:
+        self.enable_compression = False
+        self.has_query = True
+        self.num_test_iterations = 10
+        super().setUp()
+
+
+class TestCaseReaderWildcardQueryZst(TestCaseReaderWildcardQueryBase):
+    """
+    Tests stream reader against zst compressed IR stream with the query that
```
Comment (davidlion) [inline tests/test_readers.py]: zstd

File: tests/test_readers.py
Comment (davidlion) [inline tests/test_readers.py:1]: Zst and Zstd pop up interchangeably, we should probably use 'Zstd' everywhere.

Comment (LinZhihao-723) [reply tests/test_readers.py:1]: Everything should be "zstd" now to be consistent

**Review by davidlion — APPROVED** [review]

**Pull request merged**